### PR TITLE
net: lib: nrf_cloud: move FOTA JSON into codec

### PR DESCRIPTION
**Pull request by jayteemo**
Move JSON manipulation from the FOTA module to the
codec module.
IRIS-6062

### CODE REVIEW

**Comment by jayteemo** [comment]
Only the last commit is relevant for this PR: `net: lib: nrf_cloud: move FOTA JSON into codec`
Other commits belong to https://github.com/nrfconnect/sdk-nrf/pull/11169

**Review by plskeggs — APPROVED** [review]

File: subsys/net/lib/nrf_cloud/include/nrf_cloud_codec_internal.h [hunk 1]
```diff
@@ -184,13 +223,21 @@ bool nrf_cloud_disconnection_request_decode(const char *const buf);
...
+int json_array_str_add(cJSON * const array, const char * const string);
+
+/** @brief Add an integer value to the provided cJSON array. */
+int json_array_num_add(cJSON * const array, const int number);
 
 /** @brief Obtain a pointer to the string of the specified key in the cJSON object.
  * No memory is allocated, pointer is valid as long as the cJSON object is valid.
  */
-int get_string_from_obj(const cJSON * const obj, const char *const key,
+int get_string_from_obj(const cJSON * const obj, const char * const key,
```
[the remark this answers is not in the document]
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/include/nrf_cloud_codec_internal.h:240]: To appease the compliance check 🥴

[the remark this answers is not in the document]
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/include/nrf_cloud_codec_internal.h:240]: But you removed spaces in similar positions in the same commit....

[the remark this answers is not in the document]
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/include/nrf_cloud_codec_internal.h:240]: the compliance checker works in mysterious ways 🤷

File: subsys/net/lib/nrf_cloud/include/nrf_cloud_codec_internal.h [hunk 2]
```diff
@@ -134,22 +136,59 @@ int nrf_cloud_dev_status_json_encode(const struct nrf_cloud_device_status *const
 /** @brief Free memory allocated by @ref nrf_cloud_shadow_dev_status_encode */
 void nrf_cloud_device_status_free(struct nrf_cloud_data *status);
 
-/** @brief Free memory allocated by @ref nrf_cloud_rest_fota_execution_decode */
+/** @brief Free memory allocated by @ref nrf_cloud_rest_fota_execution_decode or
+ * @ref nrf_cloud_fota_job_decode
+ */
 void nrf_cloud_fota_job_free(struct nrf_cloud_fota_job_info *const job);
 
+/** @brief Create an nRF Cloud MQTT FOTA job update payload which is to be sent
```
Comment (peknis) [inline subsys/net/lib/nrf_cloud/include/nrf_cloud_codec_internal.h:144]: Five attributes to the payload. May be hard to split, but still remarkable achievement ;).

File: subsys/net/lib/nrf_cloud/include/nrf_cloud_codec_internal.h [hunk 3]
```diff
@@ -134,22 +136,59 @@ int nrf_cloud_dev_status_json_encode(const struct nrf_cloud_device_status *const
...
+ * on the FOTA jobs update topic.
+ * If successful, memory is allocated for the provided object.
+ * The @ref nrf_cloud_obj_free function should be called when finished with the object.
+ */
+int nrf_cloud_obj_fota_job_update_create(struct nrf_cloud_obj *const obj,
+					 const struct nrf_cloud_fota_job *const update);
+
+#if defined(CONFIG_NRF_CLOUD_FOTA_BLE_DEVICES)
+/** @brief Create an nRF Cloud MQTT BLE FOTA job request payload which is to be sent
```
Comment (peknis) [inline subsys/net/lib/nrf_cloud/include/nrf_cloud_codec_internal.h:153]: And this one has even one more.

**Review by peknis — APPROVED** [review]

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c [hunk 4]
```diff
@@ -1058,57 +905,29 @@ static int publish(const struct mqtt_publish_param *const pub)
...
-
-	int ret;
-	char *array_str = cJSON_PrintUnformatted(array);
-
-	cJSON_Delete(array);
-
-	if (array_str == NULL) {
-		return -ENOMEM;
-	}
+static int publish_and_free(struct nrf_cloud_obj *const pub_data,
```
Comment (glarsennordic) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:908]: Maybe `publish_and_free_message`?

**Review by glarsennordic — APPROVED** [review]
Looks great, everything here is either nits or suggestions

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c [hunk 5]
```diff
@@ -1426,34 +1234,24 @@ int nrf_cloud_fota_ble_update_check(const bt_addr_t *const ble_id)
...
-	if (!array) {
-		return -ENOMEM;
-	}
-
-	if (!add_string_to_array(array, ble_id_str)) {
-		cJSON_Delete(array);
-		return -ENOMEM;
+	ret = nrf_cloud_obj_fota_ble_job_request_create(&check_obj, ble_id);
+	if (ret) {
+		LOG_ERR("Failed to create BLE FOTA job update check, error: %d", ret);
```
[the remark this answers is not in the document]
Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:1247]: Ahh, gotcha, makes sense then

**Pull request merged**